algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -21,249 +20,286 @@ use crate::{
             .interpolate_with_pc(ifft_precomputation);
-        end_timer!(f_poly_time);
-        let g = DensePolynomial::from_coefficients_slice(&f.coeffs[1..]);
-        let h = &a_poly
-            - &{
-                let mut multiplier = PolyMultiplier::new();
-                multiplier.add_polynomial_ref(&b_poly, "b");
-                multiplier.add_polynomial_ref(&f, "f");
-                multiplier.add_precomputation(fft_precomputation, ifft_precomputation);
-                multiplier.multiply().unwrap()
-            };
-        // Let K_max = largest_non_zero_domain;
-        // Let K = non_zero_domain;
-        // Let s := K_max.selector_polynomial(K) = (v_K_max / v_K) * (K.size() / K_max.size());
-        // Let v_K_max := K_max.vanishing_polynomial();
-        // Let v_K := K.vanishing_polynomial();
-        // Let lhs := h / v_K * (K.size() / K_max.size());
-
-        // Later on, we multiply `h` by s, and divide by v_K.
-        // Substituting in s, we get that h * s / v_K_max = h / v_K * (K.size() / K_max.size());
+        let mut multiplier = PolyMultiplier::new();
+        multiplier.add_precomputation(fft_precomputation, ifft_precomputation);
+        multiplier.add_polynomial(m_at_alpha, "m_at_alpha");
+        multiplier.add_polynomial(assignment, "assignment");
+        let mut z_m_at_alpha = multiplier.multiply().unwrap();
+        let sum = z_m_at_alpha.evaluate_over_domain_by_ref(*variable_domain).evaluations.into_iter().sum::<F>();
+        end_timer!(z_m_at_alpha_time);
+
+        // Let C = largest_domain;
+        // Let C_i = domain;
+        // Let v_C := C.vanishing_polynomial();
+        // Let v_C_i := C_i.vanishing_polynomial();
+        // Let s_i := C.selector_polynomial(C_i) = (v_C / v_C_i) * (C_i.size() / C.size());
+        // Let m := masking polynomial
+        // Let c_i := circuit combiner
+        // Let lhs_i := circuit_specific partial left hand size of the sumcheck, see protocol docs for more details
+
+        // Instead of just multiplying each lhs_i by `s_i*c_i`, we reorder the sumcheck to cancel out some terms
+        // This removes a mul and div by v_C operation over each circuit's (max_variable_domain - variable_domain)
+
+        // Later on, we divide each `lhs_i` and m by v_C, then multiply by `s_i*c_i` and sum them.
+        // Substituting in s, we get that:
+        // (\sum_i{lhs_i} + m)/v_C = \sum{h_1_i*v_C + x_g_1_i}
+        // \sum_i{c_i*s_i*(lhs_i/v_C - x_g_1_i)} + (m/v_C - x_g_1_m) = \sum{h_1_i*v_C}
+        // \sum_i{c_i*(C_i.size()/C.size())*(lhs_i/v_C_i - x_g_1_i*v_C/v_C_i)} + (m/v_C - x_g_1_m) = \sum{h_1_i*v_C}
+        // \sum_i{c_i*(C_i.size()/C.size())*(lhs_i/v_C_i} + m/v_C = \sum{h_1_i*v_C} + \sum{c_i*x_g_1_i*(v_C/v_C_i)*(C_i.size()/C.size())} + x_g_1_m
+        // (\sum_i{c_i*s_i*lhs_i} + m)/v_C = \sum{h_1_i*v_C} + \sum{c_i*s_i*x_g_1_i} + x_g_1_m
+        // (\sum_i{c_i*s_i*lhs_i} + m)/v_C = h_1*v_C + x_g_1
```

Comment:
I think by now we're repeating this logic 3 times, once in round 1, once here, and then in round 5. Maybe it's worth abstracting out at this point

Reply:
Good idea, they are in round 2, 3 and 4, took a stab at it: d6d5f11d3
Just FYI there are two funky differences:
- in round 3, we also want each circuit poly's remainder to add up, this is what the `remainder_witness` flag is for
- in round 4, we are not applying randomizers yet as we're getting those from the Verifier.